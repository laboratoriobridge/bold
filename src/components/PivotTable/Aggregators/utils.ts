import { Aggregator, AggregatorEnum, AggregatorLabels } from './model'

export const getCountAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.COUNT,
  label: aggregatorLabels.count,
  value: undefined,
  keyDependent: false,
})

const getPercentageAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.PERCENTAGE,
  label: aggregatorLabels.percentage,
  value: (values: number[], total: number): number => (values.reduce((prev, curr) => prev + curr, 0) * 100) / total,
  keyDependent: false,
  chain: [getCountAggregator(aggregatorLabels)],
  suffix: '%',
})

const getAverageAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.AVERAGE,
  label: aggregatorLabels.average,
  value: (values: number[]): number => values.reduce((prev, curr) => prev + curr, 0) / values.length,
  keyDependent: true,
})

const getMaximumAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.MAXIMUM,
  label: aggregatorLabels.maximum,
  value: (values: number[]): number => values.reduce((prev, curr) => (prev > curr ? prev : curr)),
  keyDependent: true,
})

const getMinimumAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.MINIMUM,
  label: aggregatorLabels.minimum,
  value: (values: number[]): number => values.reduce((prev, curr) => (prev < curr ? prev : curr)),
  keyDependent: true,
})

export const getAggregators = (aggregatorLabels: AggregatorLabels) => {
  return [
    getCountAggregator(aggregatorLabels),
    getPercentageAggregator(aggregatorLabels),
    getAverageAggregator(aggregatorLabels),
    getMaximumAggregator(aggregatorLabels),
    getMinimumAggregator(aggregatorLabels),
  ]
}

export const getKeyNotDependentAggregators = (aggregators: Aggregator[]) => {
  return aggregators.filter((agg) => !agg.keyDependent)
}
