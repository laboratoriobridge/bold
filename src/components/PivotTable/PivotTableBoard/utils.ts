import { difference, intersection } from 'lodash'

import { BoardField, FieldFiltersByKey, FieldValuesByKey } from './model'

interface InitialKeyPositionsAndValues<T> {
  initialRowKeys: Array<keyof T>
  initialColumnKeys: Array<keyof T>
  initialAvailableKeys: Array<keyof T>
}

export function getInitialKeys<T>(
  keys: FieldValuesByKey<T>,
  initialFields?: Array<BoardField<T>>
): InitialKeyPositionsAndValues<T> {
  const rowKeys: Array<keyof T> = []
  const columnKeys: Array<keyof T> = []

  initialFields?.forEach(({ key, origin }) => {
    if (origin === 'row') {
      rowKeys.push(key)
    }
    if (origin === 'column') {
      columnKeys.push(key)
    }
  })
  const availableKeys: Array<keyof T> = difference(Array.from(keys.keys()), [...rowKeys, ...columnKeys])

  return {
    initialRowKeys: rowKeys,
    initialColumnKeys: columnKeys,
    initialAvailableKeys: availableKeys,
  }
}

export function initializeActiveFilters<T>(
  allFiltersByKey: FieldFiltersByKey<T>,
  keys: FieldValuesByKey<T>,
  initialFields?: Array<BoardField<T>>,
  setFilterState?: (filter: FieldFiltersByKey<T>) => void
): void {
  const activeFiltersByKey = new Map(allFiltersByKey)

  initialFields?.forEach(({ key, filters }) => {
    if (filters && filters.length > 0) {
      const currentFieldValues = keys.get(key)
      const validFilters = intersection(filters, currentFieldValues)
      activeFiltersByKey.set(key, new Set(validFilters))
    }
  })

  setFilterState?.(activeFiltersByKey)
}

export function handleTagFilterRemove<T extends object>(
  key: keyof T,
  value: string,
  filterState: FieldFiltersByKey<T>
): Set<string> {
  const values = Array.from(filterState.get(key) ?? new Set<string>())

  return new Set(values.filter((item) => item !== value))
}
