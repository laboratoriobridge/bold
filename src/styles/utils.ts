import { ColorScale } from './colors'

export const hexToRGB = (hex: string, alpha?: number) => {
  const r = parseInt(hex.slice(1, 3), 16)
  const g = parseInt(hex.slice(3, 5), 16)
  const b = parseInt(hex.slice(5, 7), 16)

  if (alpha) {
    return `rgba(${r}, ${g}, ${b}, ${alpha})`
  } else {
    return `rgb(${r}, ${g}, ${b})`
  }
}

export const invertColorScale = (scale: ColorScale): ColorScale => {
  return {
    c10: scale.c100,
    c20: scale.c90,
    c30: scale.c80,
    c40: scale.c70,
    c50: scale.c60,
    c60: scale.c50,
    c70: scale.c40,
    c80: scale.c30,
    c90: scale.c20,
    c100: scale.c10,
  }
}
