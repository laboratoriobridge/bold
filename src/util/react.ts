import { MutableRefObject, Ref } from 'react'

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}
