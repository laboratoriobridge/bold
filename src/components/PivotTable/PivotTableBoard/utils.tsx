import { difference, intersection } from 'lodash'

import { ReactElement } from 'react'
import React from 'react'
import { css } from 'emotion'
import { PivotTableProps } from '../PivotTableGrid/model'
import { Tag } from '../../Tag'
import { HFlow } from '../../HFlow'
import { blue } from '../../../styles/colors'
import { Heading } from '../../Heading'
import { Icon } from '../../Icon'
import { modal } from '../../Modal'
import { LocaleConfiguration } from '../../../i18n'
import { BoardField, FieldValuesByKey } from './model'

interface InitialKeyPositionsAndValues<T> {
  initialRowKeys: Array<keyof T>
  initialColumnKeys: Array<keyof T>
  initialAvailableKeys: Array<keyof T>
}

export function getInitialKeysAndFilters<T>(
  valuesByKey: FieldValuesByKey<T>,
  allFiltersByKey: Map<keyof T, Set<string>>,
  initialFields?: Array<BoardField<T>>,
  setFilterState?: (filter: Map<keyof T, Set<string>>) => void
): InitialKeyPositionsAndValues<T> {
  const rowKeys: Array<keyof T> = []
  const columnKeys: Array<keyof T> = []
  const filtersByKey = new Map(allFiltersByKey)

  initialFields?.forEach(({ key, origin, filters }) => {
    if (origin === 'row') rowKeys.push(key)
    if (origin === 'column') columnKeys.push(key)

    if (filters.length > 0) {
      const validFilters = intersection(filters, valuesByKey.get(key))
      filtersByKey.set(key, new Set(validFilters))
    }
  })

  const availableKeys: Array<keyof T> = difference(Array.from(valuesByKey.keys()), [...rowKeys, ...columnKeys])

  setFilterState?.(filtersByKey)

  return {
    initialRowKeys: rowKeys,
    initialColumnKeys: columnKeys,
    initialAvailableKeys: availableKeys,
  }
}

export const renderClearTableModal = (labels: LocaleConfiguration['pivotTableBoard'], onTableReset: () => void) => () =>
  modal({
    size: 'small',
    render: () => (
      <Heading level={3}>
        <div css={styles.titleWrapper}>
          <Icon icon={'exclamationTriangleOutline'} fill={'danger'} size={3} style={styles.icon} />
          {labels.clearTableTitle}
        </div>
      </Heading>
    ),
    actions: [
      { label: labels.clearTableCancel },
      { label: labels.clearTableConfirm, kind: 'danger', onClick: onTableReset },
    ],
  })()

function handleTagFilterRemove<T extends object>(
  key: keyof T,
  value: string,
  filterState: Map<keyof T, Set<string>>,
  handleFilterUpdate: (key: keyof T, values: Set<string>) => void
) {
  const values = filterState.get(key) || new Set<string>()
  values.delete(value)
  handleFilterUpdate(key, values)
}

export function getFilterValuesTags<T extends object>(
  filterState: Map<keyof T, Set<string>>,
  keys: FieldValuesByKey<T>,
  keyMapping: PivotTableProps<T>['keysMapping'],
  handleFilterUpdate: (key: keyof T, values: Set<string>) => void
) {
  const filterValuesTags: ReactElement[] = []

  for (let [key, values] of Array.from(filterState)) {
    const tags: ReactElement[] = []

    if (keys.get(key)?.length === values.size || values.size === 0) {
      continue
    }

    let idx = 0
    for (let value of Array.from(values)) {
      if (idx < 3) {
        tags.push(
          <Tag
            key={value}
            removable
            onRemove={() => handleTagFilterRemove(key, value, filterState, handleFilterUpdate)}
            style={styles.tag}
          >
            {keyMapping.get(key).formatter?.(value) ?? value}
          </Tag>
        )
      } else {
        break
      }
      idx++
    }

    if (values.size > 3) {
      tags.push(
        <Tag key={key as string} type='info' style={styles.tag}>
          {`+ ${values.size - 3} ${keyMapping.get(key).keyName}`}
        </Tag>
      )
    }

    filterValuesTags.push(
      <HFlow hSpacing={0.25} alignItems='center' key={key as string}>
        <div>{`${keyMapping.get(key).keyName}`}</div>
        <div css={styles.tagsContainer}>{tags}</div>
      </HFlow>
    )
  }
  return filterValuesTags
}

const styles = {
  tag: css`
    background-color: ${blue.c90};
    color: ${blue.c40};
    border: solid 1px ${blue.c40};
    margin-left: 0.5rem;
    margin-bottom: 0.25rem;
  `,
  tagsContainer: css`
    display: flex;
    flex-wrap: wrap;
  `,
  titleWrapper: css`
    display: flex;
    align-items: center;
  `,
  icon: css`
    margin-right: 1rem;
    vertical-align: middle;
  `,
}
