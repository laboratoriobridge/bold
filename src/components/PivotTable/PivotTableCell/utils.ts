import { Theme } from '../../../styles'
import { clamp } from '../../../util'

interface CellColorProps {
  color: string
  backgroundColor: string
}

const MIN_COLOR_INDEX_OF_LIGHT_BG = 70

/**
 * Calculates text and background colors based on cell type and its content.
 *
 * If cell type is only 'value' and its content is a number:
 * - increase saturation of background color every 10% of maximum value
 * - makes the text color lighter when background becames too dark
 *
 * Otherwise, keep the default background and text color.
 *
 * @param theme
 * @param isOnlyValue True if cell type includes only 'value'
 * @param maxValue The maximum expected value for all the table
 * @param cellContent The cell's content
 * @returns An object containing the text and background colors
 */
export const calculateCellColor = (
  theme: Theme,
  isOnlyValue: boolean,
  maxValue: number,
  cellContent?: number | string
): CellColorProps => {
  let color = theme.pallete.gray.c10
  let backgroundColor = theme.pallete.primary.c100
  const cellValue = Number(cellContent)

  if (isOnlyValue && cellValue) {
    const percentageOfMaxValue = steppedPercentage(cellValue, maxValue, 10)
    const bgColorIndex = clamp(110 - percentageOfMaxValue, 10, 100)

    if (bgColorIndex < MIN_COLOR_INDEX_OF_LIGHT_BG) {
      color = theme.pallete.gray.c100
    }

    backgroundColor = theme.pallete.primary[`c${bgColorIndex}`]
  }

  return {
    color,
    backgroundColor,
  }
}

const steppedPercentage = (value: number, baseValue: number, step: number) =>
  Math.ceil(((value / baseValue) * 100) / step) * step

/**
 * Selects all the elements that has the given row and column numbers as its 'data-rownumber' and 'data-columnnumber' attributes
 */
export const selectPivotTableCellElements = (row: number, column: number): NodeListOf<HTMLElement> =>
  document.querySelectorAll(`div[data-rownumber~="${row}"], div[data-columnnumber~="${column}"]`)
