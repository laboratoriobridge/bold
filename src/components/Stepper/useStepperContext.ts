import { createContext, useContext } from 'react'
import { StepperDirection } from './Stepper'
import { StepStatus } from './Step'

export interface StepperContextValue {
  direction: StepperDirection
  gap: number | undefined
  getNextStepStatus: (currentIndex: number) => StepStatus | undefined
  incrementStep: () => number
}

const StepperContext = createContext<StepperContextValue | undefined>(undefined)

export const StepperContextProvider = StepperContext.Provider

export function useStepperContext() {
  const context = useContext(StepperContext)

  if (!context) {
    throw new Error(
      'Stepper subcomponents (like Step, StepLabel, StepConnector and StepContent) must be used inside <Stepper>'
    )
  }

  return context
}
