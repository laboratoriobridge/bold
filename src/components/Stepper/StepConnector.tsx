import React, { CSSProperties } from 'react'

import { Theme, useStyles } from '../../styles'
import { isNil } from '../../util'
import { StepStatus } from './Step'
import { StepperDirection } from './Stepper'
import { useStepperContext } from './useStepperContext'

interface StepConnectorProps {
  status: StepStatus
  direction: StepperDirection
  labelHeight: number
}

export function StepConnector(props: StepConnectorProps) {
  const { status, direction, labelHeight } = props

  const { gap } = useStepperContext()
  const { classes, css } = useStyles(createStyles, status, gap, labelHeight)

  return (
    <span
      data-testid='step-connector'
      className={css(
        classes.baseConnector,
        direction === 'horizontal' ? classes.connectorHorizontal : classes.connectorVertical
      )}
    />
  )
}

const createStyles = (theme: Theme, status: StepStatus, gap: number | undefined, labelHeight: number) => {
  const hasGap = !isNil(gap)
  const borderStyle = status === 'inactive' ? 'dashed' : 'solid'
  const borderColor =
    status === 'incompleted' || status === 'inactive' ? theme.pallete.gray.c80 : theme.pallete.primary.main

  return {
    baseConnector: {
      position: 'absolute',
      transition: 'border-color .4s ease',
    } as CSSProperties,
    connectorHorizontal: {
      width: hasGap ? `calc(100% + ${gap}rem - 1rem)` : 'calc(100% - 1rem)',
      top: 'calc(0.75rem - 1px)',
      left: 'calc(50% + 0.5rem)',
      borderTopWidth: '2px',
      borderTopStyle: borderStyle,
      borderTopColor: borderColor,
    } as CSSProperties,
    connectorVertical: {
      height: hasGap ? `calc(100% + ${gap}rem - 1rem)` : 'calc(100% - 1rem)',
      top: `calc(${labelHeight}px / 2 + 0.5rem)`,
      left: 'calc(0.75rem - 1px)',
      borderRightWidth: '2px',
      borderRightStyle: borderStyle,
      borderRightColor: borderColor,
    } as CSSProperties,
  }
}
