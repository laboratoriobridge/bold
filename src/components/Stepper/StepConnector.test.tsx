import { render } from '@testing-library/react'
import React from 'react'
import { createMockStepperContext } from '../../test/utils/createMockStepperContext'
import { StepConnector } from './StepConnector'
import { StepperContextProvider } from './useStepperContext'

it('should render correctly', () => {
  const mockContextValue = createMockStepperContext()

  const { container } = render(
    <StepperContextProvider value={mockContextValue}>
      <StepConnector status='active' direction='horizontal' labelHeight={10} />
    </StepperContextProvider>
  )
  expect(container).toMatchInlineSnapshot(`
    .emotion-0 {
      position: absolute;
      -webkit-transition: border-color .4s ease;
      transition: border-color .4s ease;
      width: calc(100% - 1rem);
      top: calc(0.75rem - 1px);
      left: calc(50% + 0.5rem);
      border-top-width: 2px;
      border-top-style: solid;
      border-top-color: #0069D0;
    }

    <div>
      <span
        class="emotion-0"
        data-testid="step-connector"
      />
    </div>
  `)
})

it('should render span with data-testid', () => {
  const mockContextValue = createMockStepperContext()

  const { getByTestId } = render(
    <StepperContextProvider value={mockContextValue}>
      <StepConnector status='active' direction='horizontal' labelHeight={10} />
    </StepperContextProvider>
  )
  expect(getByTestId('step-connector')).toBeInTheDocument()
})

it('should apply correct styles when direction is horizontal', () => {
  const mockContextValue = createMockStepperContext()

  const { getByTestId } = render(
    <StepperContextProvider value={mockContextValue}>
      <StepConnector status='active' direction='horizontal' labelHeight={10} />
    </StepperContextProvider>
  )

  const connector = getByTestId('step-connector')

  expect(getComputedStyle(connector).width).toBe('calc(100% - 1rem)')
  expect(getComputedStyle(connector).height).toBe('')
  expect(getComputedStyle(connector).borderTopWidth).toBe('2px')
  expect(getComputedStyle(connector).borderRightWidth).toBe('')
  expect(getComputedStyle(connector).top).toBe('calc(0.75rem - 1px)')
  expect(getComputedStyle(connector).left).toBe('calc(50% + 0.5rem)')
})

it('should apply correct styles when direction is vertical', () => {
  const mockContextValue = createMockStepperContext()

  const { getByTestId } = render(
    <StepperContextProvider value={mockContextValue}>
      <StepConnector status='active' direction='vertical' labelHeight={10} />
    </StepperContextProvider>
  )

  const connector = getByTestId('step-connector')

  expect(getComputedStyle(connector).height).toBe('calc(100% - 1rem)')
  expect(getComputedStyle(connector).width).toBe('')
  expect(getComputedStyle(connector).borderRightWidth).toBe('2px')
  expect(getComputedStyle(connector).borderTopWidth).toBe('')
  expect(getComputedStyle(connector).top).toBe('calc(10px / 2 + 0.5rem)')
  expect(getComputedStyle(connector).left).toBe('calc(0.75rem - 1px)')
})

it('should use gap from context when defined', () => {
  const gap = 2
  const mockContextValue = createMockStepperContext({ gap })

  const { getByTestId } = render(
    <StepperContextProvider value={mockContextValue}>
      <StepConnector status='active' direction='horizontal' labelHeight={10} />
    </StepperContextProvider>
  )

  const connector = getByTestId('step-connector')

  expect(getComputedStyle(connector).width).toBe(`calc(100% + ${gap}rem - 1rem)`)
})

it('should change border when status is active', () => {
  const mockContextValue = createMockStepperContext()

  const { getByTestId } = render(
    <StepperContextProvider value={mockContextValue}>
      <StepConnector status='active' direction='horizontal' labelHeight={10} />
    </StepperContextProvider>
  )

  const connector = getByTestId('step-connector')

  expect(getComputedStyle(connector).borderTopStyle).toBe('solid')
  expect(getComputedStyle(connector).borderTopColor).toBe('#0069D0')
})

it('should change border when status is completed', () => {
  const mockContextValue = createMockStepperContext()

  const { getByTestId } = render(
    <StepperContextProvider value={mockContextValue}>
      <StepConnector status='completed' direction='horizontal' labelHeight={10} />
    </StepperContextProvider>
  )

  const connector = getByTestId('step-connector')

  expect(getComputedStyle(connector).borderTopStyle).toBe('solid')
  expect(getComputedStyle(connector).borderTopColor).toBe('#0069D0')
})

it('should change border when status is incompleted', () => {
  const mockContextValue = createMockStepperContext()

  const { getByTestId } = render(
    <StepperContextProvider value={mockContextValue}>
      <StepConnector status='incompleted' direction='horizontal' labelHeight={10} />
    </StepperContextProvider>
  )

  const connector = getByTestId('step-connector')

  expect(getComputedStyle(connector).borderTopStyle).toBe('solid')
  expect(getComputedStyle(connector).borderTopColor).toBe('#D3D4DD')
})

it('should change border when status is inactive', () => {
  const mockContextValue = createMockStepperContext()

  const { getByTestId } = render(
    <StepperContextProvider value={mockContextValue}>
      <StepConnector status='inactive' direction='horizontal' labelHeight={10} />
    </StepperContextProvider>
  )

  const connector = getByTestId('step-connector')

  expect(getComputedStyle(connector).borderTopStyle).toBe('dashed')
  expect(getComputedStyle(connector).borderTopColor).toBe('#D3D4DD')
})
