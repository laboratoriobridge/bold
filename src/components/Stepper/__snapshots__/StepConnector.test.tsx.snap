// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
.emotion-0 {
  position: absolute;
  -webkit-transition: border-color .4s ease;
  transition: border-color .4s ease;
  width: calc(100% - 1rem);
  top: calc(0.75rem - 1px);
  left: calc(50% + 0.5rem);
  border-top-width: 2px;
  border-top-style: solid;
  border-top-color: #0069D0;
}

<div>
  <span
    class="emotion-0"
    data-testid="step-connector"
  />
</div>
`;
